scalyr_agent/agent_main.py

```diff
@@ -757,40 +757,54 @@ def __detailed_status(
             return 1
 
         return_code = 0
-        fp = open(status_file)
-        for line in fp:
-            if not health_check:
-                print(line.rstrip())
-
-            if status_format == "json" or health_check:
-                health_result = self.__find_health_result_in_status_json(line)
-                if health_result:
-                    if health_check:
-                        print("Health check: %s" % health_result)
-                    if health_result != "Good":
-                        return_code = 2
-                elif health_check:
-                    print("Cannot get health check result.")
-            elif (
-                status_format == "text"
-                and "Health check:" in line
-                and not re.match(r"^Health check\:\s+Good$", line.strip())
-            ):
-                return_code = 2
-        fp.close()
+
+        with open(status_file, "r") as fp:
+            content = fp.read()
+
+        if not health_check:
+            # Regular non-health check invocation, just print the stats
+            print(content)
+            return return_code
```

Comment:
I believe you aren't covering a case here.  We will return `0` for `return_code` if the status format is `json` and `health_check` is not True.  That's a change in behavior.  We are always trying to return a status code based on whether or not the agent is healthy.

Reply:
Ah, that's a good catch - I forgot we want to do that (and it doesn't appear we have any tests for it either).

I will make that change and also look into adding some tests to avoid such accidental regressions in the future.

I will probably start with a simple mock based unit tests, but once I have some more time, I will work on end to end tests (those will take more work since some of those things will be a bit hard to test in the end to end context).

Reply:
Here we go - d66c36a6f7cb57d67a2498c74a1eb02379da21fb.

Now there are also test cases which cover all the possible scenarios (which are quite many) - b7c01042b511492dea3c2ffe0b8f5eedb9793e00.

As mentioned above, those are not full blown end to end tests (those would be better since we would assert on the actual exit code, but that would require more work and it's on TODO for future improvements). It's still better than what we had before though (no tests for those scenarios).

A couple of things worth noting:

1. Added new argument to the method which is only used for testing purposes. I almost always try to avoid that (especially in dyamic languages like Python where it's usually easy to patch thing), but this is a bit special case. We could patch ``open`` method, but it would get much more messy and require much more glue code since we perform many different operations on the open call and open is also used as context manager...

2. We now return 3 if case we can't get health check data and -H flag is used - imo that's desired and better than previous behavior when we would return 0 in such case - https://github.com/scalyr/scalyr-agent-2/blob/master/scalyr_agent/agent_main.py#L772. This scenario should only really happen if status file for some reason is missing status data (aka running old version of the agent with new version of status command or similar - unlikely). We exit with 0 if we can't get health check data is -H flag is not used (same as before)

3. Those tests are intentionally a bit on the "copy paste" side - I could reduce some code by changing it a bit and using a loop. But in the tests, I tend to be careful - many times I've seen broken / invalid tests, especially "dynamic" ones which utilizes loops and similar.